prusti-viper/src/encoder/name_interner.rs

```diff
@@ -4,60 +4,130 @@
-    pub fn intern<S: AsRef<str>>(&mut self, full_unique_name: S, readable_names: &[S]) -> String {
-        let full_unique_name = full_unique_name.as_ref();
-
-        debug_assert!(!readable_names.iter().any(|r| r.as_ref() == ""));
-
-        // Check that the readable name is not equal to full names passed in the past.
-        debug_assert!(!readable_names.iter().any(
-            |r| r.as_ref() != full_unique_name && self.name_to_symbol.contains_key(r.as_ref())
-        ));
-
-        // Check that readable names passed in the past are not equal to the current full name.
-        debug_assert!(
-            self.name_to_symbol.contains_key(full_unique_name)
-            || !self.used_symbols.contains(full_unique_name)
-        );
-
-        // Return the symbol, if we already interned the full name
-        if let Some(symbol) = self.name_to_symbol.get(full_unique_name) {
-            debug!("Interning of {:?} is {:?}", full_unique_name, symbol);
-            return symbol.clone();
+    pub fn intern(&mut self, prefix: &str, def_id: DefId, tcx: TyCtxt) -> String {
+        let def_path = tcx.def_path(def_id);
+        // The `to_string` gives each `DisambiguatedDefPathData` separated by "::"
+        let crate_name = tcx.crate_name(def_path.krate).to_string();
+        let item_name = def_path.to_string_no_crate_verbose();
+        // Assume that the friendly `tcx.opt_item_name` is the last one, use second-to-last as the mod unique, and others as crate unique.
+        let mut name_split = item_name.rsplitn(3, "::");
+        let short_name = name_split.next().unwrap_or_default().to_string();
+        let mod_unique = name_split.next().unwrap_or_default().to_string();
+        // TODO: How does this work when we have a crate with multiple files
+        let crate_unique = name_split.next().unwrap_or_default().to_string();
+
+        let path_hash = tcx.def_path_hash(def_id);
+        //let key = NameHash::new(path_hash.stable_crate_id().to_u64(), path_hash.local_hash());
+        let key = path_hash.stable_crate_id().to_u64().to_string() + &path_hash.local_hash().to_string();
+        if !self.name_to_short.contains_key(&key) {
+            let name = self.intern_name(prefix.to_string(), crate_name, crate_unique, mod_unique, short_name);
+            self.name_to_short.insert(key.clone(), name);
         }
+        key
+    }
 
-        let symbol = readable_names.iter()
-            .find(|r| !self.used_symbols.contains(r.as_ref()))
-            .map(|r| r.as_ref())
-            .unwrap_or(full_unique_name);
-        debug!("Interning of {:?} is {:?}", full_unique_name, symbol);
-        self.used_symbols.insert(symbol.to_string());
-        self.name_to_symbol.insert(full_unique_name.to_string(), symbol.to_string());
-
-        symbol.to_string()
+    fn intern_name(&mut self, prefix: String, crate_name: String, crate_unique: String, mod_unique: String, short_name: String) -> ShortName {
```

Comment:
Is this signature going to change before the PR is ready? I think it's confusing that the method returns a `ShortName` that might be invalidated by future `intern_name` calls.